### PR TITLE
authentication design document.

### PR DESCRIPTION
**Pull request by bmccandless**
This version of the document is a translation from the google doc version.
Some light reformattingh has been applied, but the content is basically identical.
I will followup with content changes to reflect the current state, and add
sections providing more details about the deployments.

### CODE REVIEW

File: rfcs/0000-rfc-process/text.md [hunk 1]
```diff
@@ -55,18 +52,17 @@ If the change is major, the following are the steps to revising an existing RFC:
...
 
-
 ## Alternatives
 
 This process was initially done on Google Docs. As mentioned earlier, the process didn't work out well for keeping track of modifications to the RFC as versioning was not easy to use. Other ideas that was considered include [IETF's Github process](https://tools.ietf.org/html/draft-ietf-git-using-github-06) as well as the [Data Coordination Platform RFC process](https://github.com/HumanCellAtlas/dcp-community/blob/master/rfcs/text/0001-rfc-process.md).
 
 ## References
 
-[0] [IETF's Github Process](https://tools.ietf.org/html/draft-ietf-git-using-github-06)
+[0][ietf's github process](https://tools.ietf.org/html/draft-ietf-git-using-github-06)
```
Comment (maniarathi) [inline rfcs/0000-rfc-process/text.md:64]: Why are these lowercased?

Comment (maniarathi) [reply rfcs/0000-rfc-process/text.md:64]: And ditto in lines below?

Comment (tihuan) [reply rfcs/0000-rfc-process/text.md:64]: Ahh sorry, looks like a bug in `prettier` 🤦 

https://github.com/prettier/prettier/issues/3835

I merged the change in https://github.com/chanzuckerberg/single-cell/pull/67/files already, but I'll fix it in a follow up PR now. Sorry!

Comment (tihuan) [reply rfcs/0000-rfc-process/text.md:64]: https://github.com/chanzuckerberg/single-cell/pull/68 PR out! I'll resolve the conflict in this PR once that one merges into `main`! Thank you!

File: rfcs/0001-authentication/text.md [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+been enabled, then the behavior is like the current cellxgene with annotations enabled. In this
+case, there are no login/logout buttons. The user id is generated from the session id, and it
+has no meaning outside of the current browser.
+
+#### Login Flow
+
+The flow is similar to “regular app” flow in the auth0 documentation. The “Web App” is the
+cellxgene server in the diagram.
+
+![Authentication Code Flow](imgs/authcodeflow.png)
```
Comment (maniarathi) [inline rfcs/0001-authentication/text.md:199]: One clarification question. In this diagram, does the "Browser" represent the cellxgene client browser or the browser that gets opened for Auth purposes?

File: rfcs/0001-authentication/text.md [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+  there after login. This is step 1 in the diagram. /login is requested when the user clicks on
+  the “login” button.
+- /logout : this endpoint is used to clear the authentication credentials. The cookie
+  containing the ID token is removed. /logout is requested when the user clicks on the “logout” button.
+- /oauth2/callback: this endpoint is accessed during step 3 of the diagram. Auth0 redirects
+  the user to this endpoint, and the request contains the authentication code and the state
+  parameter sent in from the /login endpoint. The cellxgene server then exchanges the
+  authentication code for an ID Token. Once complete, the cellxgene server redirects the client
+  back to the dataset, and directs the browser to store the ID Token in a cookie.
+
```
Comment (maniarathi) [inline rfcs/0001-authentication/text.md:274]: I remember that the health/ API endoint ended up changing as well. Can you remark here (or maybe elsewhere is appropriate) why this change was needed and how it changed?

File: rfcs/0001-authentication/text.md [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+### Authentication in Data Portal
+
+When a user visits the Data Portal site, we determine if the user is authenticated or not by
+querying a backend endpoint (e.g. api.cellxgene.cziscience.com/corpora/is-authenticated). This
```
Comment (maniarathi) [inline rfcs/0001-authentication/text.md:278]: Was this built?

File: rfcs/0001-authentication/text.md [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+If authenticated, a Logout button may appear near the top of the page.
+
+If unauthenticated, a Login button may appear near the top of the page, with the name of the user.
+
+In a previous version of this design doc, the authentication strategy for Data Portal was to use
+the [Authentication Code Flow with Proof Key for Code Exchange (PKCE)](https://auth0.com/docs/flows/concepts/auth-code-pkce),
+implemented in the client side. This has now been changed to
+using the Authentication Code flow in the backend (same as cellxgene). This approach is
+simpler with respects to the single sign on strategy. It avoids a step where the client
+needs to access the server to set the cookie containing the tokens. It is also more secure.
```
Comment (maniarathi) [inline rfcs/0001-authentication/text.md:290]: Why is it more secure?

File: rfcs/0001-authentication/text.md [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+
+The Data Portal front end requires backend support for the following endpoints
+
+- Login: start the authentication flow, set the cookie when complete
+- Logout: logout, remove the cookie
+- Is-authenticated: check if authenticated, and return user information
```
Comment (maniarathi) [inline rfcs/0001-authentication/text.md:296]: Sorry, I can't remember if this was built or not, but can you replace with the actual endpoint name if it was?

File: rfcs/0001-authentication/text.md [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+General
+
+1. [G1] GDPR requirements: For gdpr, the main thing we need to log is user id + time stamp for
+   when they first agreed to our terms of service. The underlying requirement is we need to log
+   evidence of consent, ie the "click" of agreement to our tos.
+1. [G2] The login experience requires that the user has consented to use cookies. It is assumed
+   that this question has already been asked, and we first verify the answer before proceeding
+   with the login process.
+
+## Detailed Design | Architecture | Implementation
```
Comment (maniarathi) [inline rfcs/0001-authentication/text.md:103]: Sorry I forget, did we add any tables to the RDS for cellxgene to keep track of users? Or do we only store their identify for annotations?

File: rfcs/0001-authentication/text.md [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+case, there are no login/logout buttons. The user id is generated from the session id, and it
+has no meaning outside of the current browser.
+
+#### Login Flow
+
+The flow is similar to “regular app” flow in the auth0 documentation. The “Web App” is the
+cellxgene server in the diagram.
+
+![Authentication Code Flow](imgs/authcodeflow.png)
+
```
Comment (maniarathi) [inline rfcs/0001-authentication/text.md:200]: If you're willing/able to, I think a sequence UML diagram that showcases the steps for the login flow you've described below would be super helpful! It would show a clear picture of when the endpoints are called, by whom, and in what order. You could even substitute it for the above diagram which I (personally) don't find to be super helpful without cross checking with the wording below.

File: rfcs/0000-rfc-process/text.md (hunk 1 above)
Comment (tihuan) [reply rfcs/0000-rfc-process/text.md:64]: @maniarathi PTAL at the PR 😄🙏 Thank you!

https://github.com/chanzuckerberg/single-cell/pull/68

File: rfcs/0001-authentication/text.md [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+Cellxgene
+
+1. [CXG1] The cellxgene client will provide an authenticated user with access
+   to features that require authentication (just user annotations for now).
+1. [CXG2] An authenticated user will have the option to logout.
+1. [CXG3] An unauthenticated user will not have access to features that require
+   authentication. They will be disabled in the app, and require login on first
+   use.
+1. [CXG4] An unauthenticated user will have an option to login
+1. [CXG5] A user must not unexpectedly lose work (state of the client) when
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply rfcs/0001-authentication/text.md:49]: The requirements were meant to be neutral to the solution.   We had a few options here.
The key word is "unexpectedly".  If they have been warned then it's not unexpected...

File: rfcs/0001-authentication/text.md [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+   they login. Normally when a cellxgene dataset is accessed, there is a page
+   reload, which presents the user with the initial state. Many authentication
+   flows use a redirect, which would also reset the cellxgene state back to the
+   initial state.
+1. [CXG6] The cellxgene server will enforce the authentication for routes that
+   require authentication by checking for the ID token, and rejecting requests
+   that lack a token or have an invalid/expired token. The signature needs to be
+   validated to ensure that the token has not been tampered with. An expired ID
+   token can be refreshed (as long as the refresh token has not expired).
+1. [CXG7] The cellxgene server will have an endpoint to check if the requestor
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply rfcs/0001-authentication/text.md:59]: The cellxgene frontend needs to know if the user is authenticated in order to show an Log In button or a Log Out button.  I can add that.  I have a new endpoint for this called "userinfo".

File: rfcs/0001-authentication/text.md [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+   is authenticated. The current (“config”) endpoint can serve this purpose by
+   having an authentication section.
+1. [CXG8] To avoid causing the ID token to expire while the user is working
+   (and thus disrupting the user’s work), the token can be automatically refreshed
+   when the ID token expires. The
+   expiration date of the Refresh token could be set to the maximum of 30 days.
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply rfcs/0001-authentication/text.md:65]: Actually, this should be indefinite now.  The refresh token is itself is refreshed when the id token expires.  So, a user can be logged in indefinitely as long as they visit the website at least once per 90 days.  I can add a comment for that.

File: rfcs/0001-authentication/text.md [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+   modules to be installed if authentication is not configured. The cellxgene
+   server may provide Auth0
+   support as a configuration option, but anything specific to the deployment
+   (client_id, client_secret, etc), must be configurable and not part of the code base.
+
+Data Portal
+
+1. [C1] An authenticated user will have the option to logout
+1. [C2] An unauthenticated user will have an option to login
+
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply rfcs/0001-authentication/text.md:87]: Yeah, portal and cellxgene work in the same way.   These should be reframed as common requirements.

File: rfcs/0001-authentication/text.md [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+Single sign on
+
+1. [S1] A user authenticated in cellxgene will automatically be authenticated by Data Portal
+1. [S2] A user authenticated in Data Portal will automatically be authenticated by cellxgene
+1. [S3] Logging out of one app, logs the user out of both apps
+
+General
+
+1. [G1] GDPR requirements: For gdpr, the main thing we need to log is user id + time stamp for
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply rfcs/0001-authentication/text.md:96]: auth0 stores this.  We can get the logs or access to the their DB if we need to.

File: rfcs/0001-authentication/text.md (hunk 4 above)
Comment (bmccandless) [reply rfcs/0001-authentication/text.md:278]: It's called "userinfo" in the implementation.   Yes, it's built.

File: rfcs/0000-rfc-process/text.md (hunk 1 above)
Comment (maniarathi) [reply rfcs/0000-rfc-process/text.md:64]: Done, thank you!

File: rfcs/0001-authentication/text.md [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+
+The popup experience may also depend on the client’s browser and security settings, which could
+lead a larger testing and support burden.
+
+So, the consensus was to go with the “regular app” approach, and strongly encourage the user to
+login before doing work.
+
+## References
+
+[0][location to the authorization code flow diagram](https://auth0.com/docs/architecture-scenarios/web-app-sso/part-1)
```
Comment (tihuan) [inline rfcs/0001-authentication/text.md:379]: @bmccandless if we want the link text not be lowercase, we'll need to add `\` before the number like below, otherwise linter will force lowercase! Thank you!

```md
## References
\[0][IETF's Github Process](https://tools.ietf.org/html/draft-ietf-git-using-github-06)
\[1][Google Developer Documentation Style Guide](https://developers.google.com/style/highlights#introduction_1)
\[2][Data Coordination Platform RFC Process](https://github.com/HumanCellAtlas/dcp-community/blob/master/rfcs/text/0001-rfc-process.md)
```

File: rfcs/0001-authentication/text.md [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+
+1. cellxgene: cellxgene.cziscience.com/&lt;urlroot>/&lt;datasetname>/...
+1. Data Portal: cellxgene.cziscience.com/
+
+Back-end:
+
+1. cellxgene: api.cellxgene.cziscience.com/cellxgene/...
+1. Data Portal: api.cellxgene.cziscience.com/corpora/
+
+CloudFront (or any http proxy) can be configured to map the cellxgene API from the “api
```
[the remark this answers is not in the document]
Comment (edulop91) [reply rfcs/0001-authentication/text.md:146]: I think it did

File: rfcs/0001-authentication/text.md [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+  suggest a name like “/oauth2/callback”. The request will contain the authorization code.
+- **[4]** cellxgene can then exchange the authorization code for an ID token by sending a back
+  channel request to Auth0. Cellxgene must know the client_secret for this step. Part of the
+  configuration and startup process for cellxgene will be to retrieve this secret. In the case
+  of CZI hosted cellxgene the client_secret will be retrieved from the AWS Secrets Manager.
+- **[5]** Auth0 will send back the ID token, as a JWT (JSON Web Token).
+- The cellxgene server then redirects (full page nav) the user back to the original dataset
+  , and directs the client to save the JWT in an httpOnly cookie on the client.
+- The JWT proves the identity of the client. Subsequent requests from the client to the
+  cellxgene server will include the cookie with the JWT. When the cellxgene client re-requests
```
Comment (edulop91) [inline rfcs/0001-authentication/text.md:234]:
```suggestion
  cellxgene server will include the cookie with the `id_token`. When the cellxgene client re-requests
```
might be worth the extra specificity wrt the oauth2/oidc protocol

File: rfcs/0001-authentication/text.md [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+  the configuration through the dataset’s “config” endpoint (&lt;server>/d/&lt;datasetname>/api
+  /v0.2/config), the configuration will now indicate that the user is authenticated. The
+  cellxgene client can also have access to the user’s name, which it may use when displaying the log information.
+  1. If the ID token is invalid, then the cellxgene server can re-initiate the authentication
+     process by going back to step [1]
+  2. The server will verify the signature of the token to ensure that it was signed by the
+     sender and not altered.
+  3. If the ID token is expired, then the server will obtain a new ID token automatically
```
Comment (edulop91) [inline rfcs/0001-authentication/text.md:242]:
```suggestion
  3. If the ID token is expired, then the client will obtain a new ID token automatically
```
I think this happens client-side?

File: rfcs/0001-authentication/text.md [hunk 18]
```diff
@@ -0,0 +1,383 @@
...
+### Cellxgene could use a client side approach such as authorization code with proof key code exchange (PKCE)
+
+The login/authentication approach described in the cellxgene section uses the more secure
+Authorization Code grant type (response_type=code). This flow requires extra steps (steps [4]
+and [5] in the diagram), where the authorization code is exchanged to an ID token by making a
+request from the cellxgene server to the Auth0 server.
+
+An alternative is to use the authorization code with proof key code exchange (PKCE). This is
+the approach you might use in a single page app where you do not want the code running in a user’s
+browser to have access to the auth0 client secret.
```
Comment (edulop91) [inline rfcs/0001-authentication/text.md:334]:
```suggestion
browser to have access to the oauth2 client secret.
```

**Comment by Bento007** [comment]
I would finish up this PR in before making additional changes from https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/631undefined
